### ISSUE TITLE
[DISCUSSION] Initial setup discussion

### ISSUE THREAD

**Issue by dabura667** [bitpay/copay#4568]
I just had to introduce 32 people to bitcoin (with Copay), and it was a disaster...

Writing down the phrase while at a party with drinks was nigh impossible... Most people gave up an opted out of receiving $6 to buy a drink because it was "too difficult"...

I just wanted to open a discussion of ways to improve onboarding without reverting to the old insecure way. (Letting them use the wallet before ensuring a backup is secured)

**Comment by dabura667** [comment]
For comparison, later on that night, recommending iPhone users breadwallet had 11/14 people install, receive, and use bitcoin. So I concluded it didn't really have to do with bitcoin itself... Just the forced writing down part.

**Comment by dabura667** [comment]
## One idea

What about offering an encrypted backup storage with BWS similar to how blockchain.info stores all the encrypted wallets for users?

We could discourage this by only allowing activation of the feature from the "Needs Backup" screen. and making it a small text link on that screen in the text. "click here to store an encrypted copy in the cloud."

There are a lot of logistics to this method which would be hard to implement and get around...

It's one idea though.

**Comment by dabura667** [comment]
## Another idea

Integrate some form of cloud storage of secrets. iCloud (I think?) backs up the keychain for iPhone? Android should have some built in password managing stuff (maybe?)

Maybe integration with 1Password or LastPass or Keepass might be interesting...

**Comment by matiu** [comment]
interesting. Iam sorry the presentation didnt work...

what about this:
on the receive tab add "learn more" link that will show some extra text, a
link "i understand the risks and i will backup later..." and if the user
click there it will show the receive address but donot set the backup flag
so she will be asked again to backup next time.

something like Chrome when you enter a flaged dangerous site: learn more =>
take me back to safety or I understand the risk links.

**Comment by dabura667** [comment]
hmm, I think this could work.

One of the main groups of people who were damaged by the old way of doing things were people who:

1. Installed without knowing anything about bitcoin.
2. Installed without someone to explain about backing up importance next to them.

etc.

The flow could be something like:

1. On the "Needs backup" screen, we put a text link below the button for backup saying "Skip Backup"
2. This takes the user to a screen which explains in concise language the risk of using a wallet without backing up. (too long, and no one will read it) And at the bottom there is two buttons, one is greyed out and can not be used. "I understand." and "Make Backup". Only "Make Backup" is enabled initially.
3. After displaying the screen for 10 seconds, enable the "I understand" button. This will trigger the user to be able to use the receive tab.
4. On BWS, store the user as "ignoreBackup": true
5. If the user every goes into Wallet details and finishes the backup phrase test then change "ignoreBackup" to false and backupComplete (?) to true.
6. For any user with notifications enabled, and ignoreBackup is true, add an extra part to notification (or make a delayed notification a few seconds later) reminding them to back up.

**Comment by dan-da** [comment]
I prefer the "allow and nag" approach.

One additional thought:   people use wallets for different purposes and at different times.  Some wallets will be created just for quick spending cash, while others are lifetime investments.  

It may be worth looking at the total value that the wallet contains (btc or fiat) and nagging based on that.  Eg, there could be nag thresholds, such as:

Under $20 worth.  No nag.
Under $100 worth.  Nag every 5 uses.
Under $1000 worth.   Nag every use.
over $1000 worth.  refuse to operate until needsBackup = false.

**Comment by JulianTosh** [comment]
#4450

TL;DR If backup flag not set, only allow the user to specify an amount to receive that is less than a safe amount ($50?).

**Comment by matiu** [comment]
@JulianTosh @dan-da it is not possible to restrict incoming payment amounts for an address.  Making only possible to use 'request specific amount' have some issues:
  1- it is still possible for the sender to send more coins to the generated address (by manually altering the amount, or reusing the code multiple times).
  2- would probably left some use cases out, like the one @dabura667 is mentioning: In a deme: "install copay and give me your address"...

**Comment by KeyJockey** [comment]
Yeah filter-and-nag based on value $ amount would be cool but if that's too complex maybe at least put a limit on it for allowing the user to skip backup only like 5 times?  

The "enable" button could show a countdown like "you can only do this x more times before you really need to complete the backup".

This would prevent a new user from just habitually skipping the backup and *never* actually getting around to doing it... until disaster strikes.

**Comment by dabura667** [comment]
> The "enable" button could show a countdown like "you can only do this x more times before you really need to complete the backup".

The only problem is that a user could copy their address once and paste it everywhere and receive 100 transactions worth 1BTC each.

Which is why I think some notification for the user every time the money has arrived without a backup.

It could be like the "Incoming Payment" or "wallet synced" notifications. A banner pops down if you have opened the app. And if you have push notifications, then send one of those as well to remind the user they need to backup.

Really the only thing you can do for the user once they've got an address is to ask them to backup before they lose its private key.

**Comment by Derrick-** [comment]
An alternative first backup method such as export to email would be helpful. Sending an encrypted export to an email is much easier than finding a pencil and paper and writing down words.

**Comment by dabura667** [comment]
> (and I consider it far more secure)

Really depends. We have no control over how secure their password is or whether they can remember the password. Not to mention that depending on the email service they use, they could be sending their encrypted wallet in plaintext over the internet... which means if they have a weak password, it would be trivial to sniff and crack.

I think the main thing about these phrases, is that you can't regain lost security... so you need to assume this phrase will have a fortune in it someday, because if you don't start there, and you go the other way around "start with a dumb wallet for just learning how to send. who cares." and you are too lazy to create a new wallet, or someone sends you a lot of money, or bitcoin's price skyrockets... then suddenly you wish your wallet was more secure.

**Comment by Derrick-** [comment]
I understand that it's going to be a case where the ultimate security of either method varies for each user. For example you also have no control over how many smartphones are spying on the piece of paper the user is writing on, or what happens to that piece of paper after the install.

When I say I consider it more secure, I mean it's more secure for me, because I have a reliable email infrastructure that is backed up and has a bus factor > 1. I mean that I am very nervous writing unencrypted words on pieces of paper, and much more comfortable emailing an encrypted password.

It has seemed to me that the decision process surrounding some specific CoPays features tends to target a specific noob demographic at the exclusion of a user demographic which is more careful and capable. While I understand the targeting of users which have a lesser skill and confidence level, I don't think it often needs to be done at the expense of more experienced users, this is what options are for.

I have previously seen fast NACK's to issues which could improve the wallet for experienced users because of a fear of new user experience ( example https://github.com/bitpay/copay/issues/4519 ). Something like an "Advanced mode" could overcome a lot of these conflicts; and while I know the primary focus of this thread is to discuss making it easier for new users to get into CoPay, I don't think the frustration of experienced users should be simply ignored.

**Comment by Derrick-** [comment]
@dabura667 

> "I think the main thing about these phrases, is that you can't regain lost security... so you need to assume this phrase will have a fortune in it someday, because if you don't start there, and you go the other way around "start with a dumb wallet for just learning how to send. who cares." and you are too lazy to create a new wallet, or someone sends you a lot of money, or bitcoin's price skyrockets... then suddenly you wish your wallet was more secure."
I do very much agree with this. I do think backup is very important. I just question the physical security of having people write words on paper.

**Comment by KeyJockey** [comment]
@Derrick 
>It has seemed to me that the decision process surrounding some specific CoPays features tends to target a specific noob demographic at the exclusion of a user demographic which is more careful and capable. While I understand the targeting of users which have a lesser skill and confidence level, I don't think it often needs to be done at the expense of more experienced users, this is what options are for.

+1000 absolutely agree on this... case in point MY big pet peeve issue re: miner fee adjustment for Advanced users

**Comment by derrick** [comment]
@KeyJockey I'm afraid you pinged the wrong @Derrick- .

**Comment by KeyJockey** [comment]
Whoops @derrick , sorry about that: typed it, didn't click the popup LOL my bad 😛  

Post updated to ping the right @Derrick-  (with the hyphen)

**Comment by dabura667** [comment]
More @Derrick- never hurt anyone.

You're welcome to chime in from an outsider perspective. :-D

**Comment by carlosfaria94** [comment]
One option is using biometrics and a PIN code to encrypt the wallet and send it to the server (for devices that supported it).

**Comment by JulianTosh** [comment]
@dabura667 

I don't know why it took me so long to realized this, but the proper solution for training a crowd of people is to use the testnet. 

1. It bypasses the backup requirement so the process of installing a usable wallet is very quick.
2. TestNet coins are worthless so you can hand them out like candy.
3. The network operates the same exact way and just as fast.

Here is a testnet faucet, but if it goes down, just google 'bitcoin testnet faucet' and you'll find more.

http://tpfaucet.appspot.com/
